ott/tools/k_means.py

```diff
@@ -0,0 +1,366 @@
+# Copyright 2022 Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import functools
+import math
+from typing import Any, Callable, NamedTuple, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+from typing_extensions import Literal
+
+from ott.core import fixed_point_loop
+from ott.geometry import costs, pointcloud
+
+__all__ = ["k_means", "KMeansOutput"]
+
+Init_t = Union[Literal["k-means++", "random"],
+               Callable[[pointcloud.PointCloud, int, jnp.ndarray], jnp.ndarray]]
+
+
+class KPPState(NamedTuple):
+  key: jnp.ndarray
+  centroids: jnp.ndarray
+  centroid_dists: jnp.ndarray
+
+
+class KMeansState(NamedTuple):
+  centroids: jnp.ndarray
+  prev_assignment: jnp.ndarray
+  assignment: jnp.ndarray
+  errors: jnp.ndarray
+  center_shift: float
+
+
+class KMeansOutput(NamedTuple):
+  """Output of the :func:`~ott.tools.k_means.k_means` algorithm.
+
+  Args:
+    centroids: Array of shape ``[k, ndim]`` containing the centroids.
+    assignment: Array of shape ``[n,]`` containing the labels.
+    converged: Whether the algorithm has converged.
+    iteration: The number of iterations run.
+    error: (Weighted) sum of squared distances from each point to its closest
+      center.
+    inner_errors: Array of shape ``[max_iterations,]`` containing the ``error``
+      at every iteration.
+  """
+  centroids: jnp.ndarray
+  assignment: jnp.ndarray
+  converged: bool
+  iteration: int
+  error: float
+  inner_errors: Optional[jnp.ndarray]
+
+  @classmethod
+  def _from_state(
+      cls,
+      state: KMeansState,
+      *,
```

Comment:
was not familiar with this syntax before! is there a simple reason to force arguments after to be keyword-only?

Reply:
I like to use them for simple arguments (ints/floats/bools) so that whenever the function is called, the values have to have a meaning, rather than being opaque, e.g., `_from_states(cls, state, tol, True)`